speculos/client.py

```diff
@@ -0,0 +1,181 @@
+from contextlib import contextmanager
+from PIL import Image, ImageChops
+from typing import Generator, Tuple
+import json
+import logging
+import requests
+import socket
+import subprocess
+import time
+
+logger = logging.getLogger("speculos-client")
+logger.setLevel(logging.INFO)
+
+
+def split_apdu(data: bytes) -> Tuple[bytes, int]:
+    assert len(data) >= 2
+    status = int.from_bytes(data[-2:], "big")
+    return data[:-2], status
+
+
+def screenshot_equal(path1: str, path2: str) -> bool:
+    """Compare two images and return True if they are equal."""
+
+    with Image.open(path1) as img1:
+        with Image.open(path2) as img2:
+            diff_img = ImageChops.difference(img1, img2)
+    return diff_img.getbbox() is None
+
+
+class ApduException(Exception):
+    def __init__(self, sw=0x6F00) -> None:
+        self.sw = sw
+
+    def __str__(self) -> str:
+        return f"Exception: invalid status 0x{self.sw:x}"
+
+
+class ApduResponse:
+    def __init__(self, response: requests.Response) -> None:
+        self.response = response
+
+    def receive(self) -> bytes:
+        assert self.response.status_code == 200
+        data, status = split_apdu(bytes.fromhex(self.response.json()["data"]))
+        if status != 0x9000:
+            raise ApduException(status)
+        return data
+
+
+class Api:
+    def __init__(self, api_url: str) -> None:
+        self.api_url = api_url
+        self.timeout = 2000
+        self.session = requests.session()
+        self.stream = self._open_stream()
+
+    def _open_stream(self) -> requests.Response:
+        stream = self.session.get(f"{self.api_url}/events?stream=true", stream=True)
+        assert stream.status_code == 200
+        return stream
+
+    def get_next_event(self):
+        line = self.stream.raw.readline()
+        event = json.loads(line)
+        assert event
+        return event
+
+    def wait_for_text_event(self, text: str) -> dict:
+        """Wait until an event containing the specified text is received."""
+
+        while True:
+            event = self.get_next_event()
+            if text in event["text"]:
+                break
+        return event
+
+    def press_and_release(self, button: str):
+        assert button in ["left", "right", "both"]
+        data = {"action": "press-and-release"}
+        with self.session.post(f"{self.api_url}/button/{button}", json=data) as response:
+            assert response.status_code == 200
```

Comment:
Using `assert` to check a condition which can happen (for example when Speculos's API server is down...) is not developer-friendly. Moreover if this happens, developers have no way of knowing what went wrong from their Python code.

Could this be replaced with raising an exception (which would mean "an HTTP error occurred: ...") instead?